src/models/history/history.ts

```diff
@@ -0,0 +1,47 @@
+import { types, IJsonPatch, SnapshotIn } from "mobx-state-tree";
+import { observable } from "mobx";
+
+export const TreePatchRecord = types.model("TreePatchRecord", {
+    tree: types.string,
+    action: types.string,
+    patches: types.frozen<ReadonlyArray<IJsonPatch>>(),
+    inversePatches: types.frozen<ReadonlyArray<IJsonPatch>>()
+})
+.views(self => ({
+    getPatches(opType: HistoryOperation) {
+        switch (opType) {
+            case HistoryOperation.Undo:
+                return self.inversePatches.slice().reverse();
+            case HistoryOperation.Redo:
+                return self.patches;
+        }
+    }
```

Comment:
This file seems to have adopted four-space tabs.

Reply:
Fixed.